src/cpu/o3/tdm.cc

```diff
@@ -0,0 +1,260 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "cpu/o3/tdm.hh"
+
+#include "cpu/o3/commit.hh"
+#include "cpu/o3/cpu.hh"
+#include "cpu/o3/decode.hh"
+#include "cpu/o3/fetch.hh"
+#include "cpu/o3/iew.hh"
+#include "cpu/o3/rename.hh"
+
+namespace gem5
+{
+
+namespace o3
+{
+TopDownStats::TopDownStats(CPU *cpu, Fetch *fetch, Rename *rename,
+                           Decode *decode, IEW *iew, Commit *commit)
+    : statistics::Group(cpu, "TopDownStats"),
+      topDownL1(cpu, fetch, rename, decode, iew, commit),
+      topDownFbL2(cpu, fetch),
+      topDownBsL2(cpu, decode, iew),
+      topDownBbL2(cpu, rename, iew),
+      topDownBbMem(cpu, rename, iew)
+{}
+
+TopDownStats::TopDownL1::TopDownL1(CPU *cpu, Fetch *fetch, Rename *rename,
+                                   Decode *decode, IEW *iew, Commit *commit)
+    : statistics::Group(cpu, "TopDownL1"),
+      ADD_STAT(frontendBound,
+               statistics::units::Rate<statistics::units::Count,
+                                       statistics::units::Count>::get(),
+               "Frontend Bound, fraction of slots lost due to frontend "
+               "undersupplying the backend"),
+      ADD_STAT(badSpeculation,
+               statistics::units::Rate<statistics::units::Count,
+                                       statistics::units::Count>::get(),
+               "Bad Speculation, fraction of slots lost due to mispeculation"),
+      ADD_STAT(backendBound,
+               statistics::units::Rate<statistics::units::Count,
+                                       statistics::units::Count>::get(),
+               "Backend Bound, fraction of slots lost due to backend resource"
+               " constraints."),
+      ADD_STAT(
+          retiring,
+          statistics::units::Rate<statistics::units::Count,
+                                  statistics::units::Count>::get(),
+          "Retiring, fraction of slots successfully retired by the backend")
+{
+    // Total Slots
+    statistics::Temp totalSlots =
+        rename->getWidth() * cpu->baseStats.numCycles;
```

Comment:
Using `numCycles` will also include `idleCycles`. I don't think we should include them.